### PR TITLE
Convert all positional arguments to keyword-only

### PR DESCRIPTION
**Pull request by higucheese**
## Motivation

This PR is based on https://github.com/optuna/optuna/issues/2911, but simply changing the order of positional arguments of `create_study` or `load_study` can confuse users. So this aims to encourage them to use keyword-only arguments first without any changes to the interface. Once users get used to keyword-only arguments, it gets much easier to align their arguments.

## Description of the changes

- Change all positional arguments of `{create,load,delete,copy}_study()` to keyword-only arguments.
- When a caller of `{create,load,delete,copy}_study()` sets values as positional arguments, the decorator `_convert_positional_args` converts them to keyword arguments according to the signature of the decorated function and sets a warning.

### CODE REVIEW

File: optuna/study/study.py
Comment (nzw0301) [inline optuna/study/study.py:53]: In my understandingm `func` assumes method or function. How about specifying the type as `Callable`?
```suggestion
def _convert_positional_args(n_positional_args: int = 0) -> Callable:
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
```

**Comment by nzw0301** [comment]
Can we add tests for this decorator? [`experimental` decorator's test](https://github.com/optuna/optuna/blob/0a7f64d9b2085295738da3bfda77e041b5992dd0/tests/test_experimental.py) might be useful.

File: optuna/study/study.py
Comment (hvy) [inline optuna/study/study.py:60]: It might be a good idea to leave a code comment (somewhere in `_convert_positional_args` ?) for maintainers that the order of arguments (although keyword-only) may not be changed, at the decorated functions.

**Comment by hvy** [comment]
Was taking a quick look at the CI failure, it seemed unrelated to the change in this PR. 🤔

File: optuna/study/study.py
Comment (nzw0301) [inline optuna/study/study.py:1070]:
```suggestion
@_convert_positional_args(n_positional_args=7)
```

Comment (higucheese) [reply optuna/study/study.py:1070]: Because `directions` is already a keyword-only argument, I think the number of arguments that can be specified by user is no more than 6.

**Comment by hvy** [comment]
The CI failures are caused by the change, sorry. Not entire understood the details but using `@overload` https://mypy.readthedocs.io/en/stable/generics.html#decorator-factories seems to fix the issue, tried verifying it locally.
(C.f. https://www.python.org/dev/peps/pep-0484/ for more about `@overload`)

**Comment by toshihikoyanase** [comment]
@nzw0301 Could you continue the review, please?

**Comment by higucheese** [comment]
In commit 7282913, the most important change is here.
```diff
-        def wrapper(*args: Any, **kwargs: Any) -> Any:
+        def wrapper(*args: Any, **kwargs: Any) -> _T:
```
The typevar `_T` is inferred as `Study` for decorated functions, `{create,load}_study` and `None` for `{delete,copy}_study`. Then, a type of a return value of a decorated `create_study` is correctly inferred as `Study` which has `ask()` attribute. It was inferred as `Optional[Study]` before this fix and mypy did not allow calling its `ask()` without checking if it was not `None`.

**Comment by higucheese** [comment]
I tried the hvy-san's suggestion using `@overload`, but it seemed that mypy failed to decide types in `_convert_positional_args` because there was no return type annotation for functions.
Like,
```
optuna/study/study.py:59: error: Function is missing a return type annotation
optuna/study/study.py:60: error: Function is missing a return type annotation
optuna/study/study.py:62: error: Function is missing a type annotation for one or more arguments
```

**Comment by higucheese** [comment]
For the remaining feedbacks, I'll respond after tomorrow.

**Comment by hvy** [comment]
Ah, that makes sense. Thanks. 🙇 
As long as this private decorator is typed to return `Callable[[Callable[..., _T]], Callable[..., _T]]`, I could in my environment omit the `_T` return annotation for `wrapper` and it still worked. Just a side note.

File: optuna/study/study.py
Comment (higucheese) [inline optuna/study/study.py:61]: To check `previous_positional_arg_names` list is valid, I add signature info. of function and compare them here. This helps developers to keep mapping previous signature to new signature, but my concern is that this is just a runtime overhead for users...

File: optuna/study/study.py
Comment (higucheese) [reply optuna/study/study.py]: Respond in 2ce7150 (add comment) and fdcf2e0 (raise TypeError), f8d64b9 (arg_names as input)

**Comment by higucheese** [comment]
I have added decorator tests. Please take a look.

File: optuna/study/study.py
Comment (hvy) [reply optuna/study/study.py:61]: Sounds good. With this logic, my understanding is that there's a runtime overhead in each invocation to the decorator function, but we only need to perform this check once, when applying the decorator, i.e. during module import? We can do this moving this assertion to an outer scope, right below `def decorator(func ...):`.

**Comment by hvy** [comment]
Left a single comment but logic seems good otherwise. 

Cc. @himkt, did you want to verify this with VSCode?

**Comment by himkt** [comment]
Screen Shot 2022-02-04 at 12 22 15
Screen Shot 2022-02-04 at 12 23 52

Given `study = create_study()` `study` is correctly recognized as `Study` (before `study` is recognized as `Any`). :tada:
Although keyword arguments is correctly displayed (I'm not sure why...), it would not be blocked.

**Comment by hvy** [comment]
> it would not be blocked.

Sorry, wasn't entire sure what you meant with "blocked" here. Would you mind explaining a bit more? 🙇

**Comment by himkt** [comment]
Sorry for the ambiguous comment @hvy.
I think it is ideal to show keyword only arguments as well (in `create_study`, `...` is displayed). But I'm not sure if it is possible to show them with decorator so I use "not be blocked".

Screen Shot 2022-02-04 at 14 51 13

**Comment by hvy** [comment]
Just got clarified by @himkt that the current behavior fine and that it's not blocked in any sense. 
I also think that the logic looks good. Just that we could address https://github.com/optuna/optuna/pull/3270/files#r798350198.

File: optuna/study/study.py
Comment (higucheese) [reply optuna/study/study.py]: Thanks! Fixed in [9e6b9ad](https://github.com/optuna/optuna/pull/3270/commits/9e6b9ad3d5be6ce74be00a87e158393b1702d3c8) and it works as intended.

**Comment by higucheese** [comment]
I have introduced `ParamSpec` which is officially available in `typing` since version 3.10 (so `typing_extensions` is necessary now) and that helps to infer the arguments of the outer function from the arguments of the inner function. Then, all arguments of XXX_study() are properly inferred like
param_spec

8510858 is a fix for argument mismatch caused by the change of XXX_study().

**Comment by higucheese** [comment]
According to https://github.com/optuna/optuna/runs/5075685882, we have to add `typing_extensions` to requirements (under `checking`?) and a check if it is available. Are there any rules to introduce new module?

**Comment by himkt** [comment]
Thank you @higucheese for investigating and improving typing stuff!

> we have to add typing_extensions to requirements (under checking?)

If we add typing_extensions to checking extras, it needs to use `optuna._imports` for lazy importing. I think it's okay to add it to `install_requires` because `typing_extensions` is official package provided by python org and [has no external dependencies](https://github.com/python/typing/blob/eeb29ad833f992ea9f2f29b32723c05586cd0aec/typing_extensions/pyproject.toml).

> Are there any rules to introduce new module?

We don't have the explicit rule for that and carefully consider for each case.
@hvy What do you think about introducing `typing_extensions` to `install_requires`?

---

log:

```
> pip install .                                                                                                                                                                                                  2022-02-05 14:45:30
Processing /Users/himkt/work/github.com/himkt/optuna
  Installing build dependencies ... done
...
Successfully installed Mako-1.1.6 MarkupSafe-2.0.1 PrettyTable-3.0.0 PyYAML-6.0 alembic-1.7.6 attrs-21.4.0 autopage-0.5.0 cliff-3.10.0 cmaes-0.8.2 cmd2-2.3.3 colorlog-6.6.0 greenlet-1.1.2 numpy-1.22.2 optuna-3.0.0a1.dev0 packaging-21.3 pbr-5.8.0 pyparsing-3.0.7 pyperclip-1.8.2 scipy-1.7.3 sqlalchemy-1.4.31 stevedore-3.5.0 tqdm-4.62.3 wcwidth-0.2.5
WARNING: You are using pip version 21.3.1; however, version 22.0.3 is available.
You should consider upgrading via the '/Users/himkt/work/github.com/himkt/optuna/venv_tiny/bin/python3 -m pip install --upgrade pip' command.
>
>
> python -c 'import optuna'                                                                                                                                                                                      2022-02-05 14:46:01
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/himkt/work/github.com/himkt/optuna/optuna/__init__.py", line 7, in <module>
    from optuna import multi_objective
  File "/Users/himkt/work/github.com/himkt/optuna/optuna/multi_objective/__init__.py", line 2, in <module>
    from optuna.multi_objective import samplers  # NOQA
  File "/Users/himkt/work/github.com/himkt/optuna/optuna/multi_objective/samplers/__init__.py", line 1, in <module>
    from optuna.multi_objective.samplers._adapter import _MultiObjectiveSamplerAdapter  # NOQA
  File "/Users/himkt/work/github.com/himkt/optuna/optuna/multi_objective/samplers/_adapter.py", line 6, in <module>
    from optuna.samplers import BaseSampler
  File "/Users/himkt/work/github.com/himkt/optuna/optuna/samplers/__init__.py", line 1, in <module>
    from optuna.samplers._base import BaseSampler
  File "/Users/himkt/work/github.com/himkt/optuna/optuna/samplers/_base.py", line 8, in <module>
    from optuna.study import Study
  File "/Users/himkt/work/github.com/himkt/optuna/optuna/study/__init__.py", line 4, in <module>
    from optuna.study.study import copy_study
  File "/Users/himkt/work/github.com/himkt/optuna/optuna/study/study.py", line 19, in <module>
    from typing_extensions import ParamSpec
ModuleNotFoundError: No module named 'typing_extensions'
```

**Comment by nzw0301** [comment]
@himkt could you jump in this PR as a reviewer instead of me?

**Comment by himkt** [comment]
Sure, assigned!

**Comment by himkt** [comment]
@higucheese Sorry for bit late. We (@himkt, @hvy, @toshihikoyanase, and @HideakiImamura) approve to add `typing_extensions` to `install_requires`. Could you please update `setup.py`?

File: setup.py
Comment (himkt) [inline setup.py:42]:
```suggestion
        "typing_extensions>=3.10.0.0",
```

Without the version constraint, `pip install .[test]` or `pip install .[integration]` seemingly install old `typing_extensions`.

Comment (higucheese) [reply setup.py:42]: Thank you for investigation! Let's try this change.

**Comment by higucheese** [comment]
The original problem caused by old `typing_extensions` seems to be fixed, but the remaining failed test seems not to be related to this PR... (`E   ModuleNotFoundError: No module named 'keras'`)

https://github.com/optuna/optuna/runs/5091139491?check_suite_focus=true
```
Run if [ 3.6 = 3.6 ]; then
============================= test session starts ==============================
platform linux -- Python 3.6.[15](https://github.com/optuna/optuna/runs/5091139491?check_suite_focus=true#step:7:15), pytest-7.0.0, pluggy-1.0.0
rootdir: /home/runner/work/optuna/optuna, configfile: pyproject.toml
collected 350 items / 1 error / 349 selected

==================================== ERRORS ====================================
____________ ERROR collecting tests/integration_tests/test_keras.py ____________
ImportError while importing test module '/home/runner/work/optuna/optuna/tests/integration_tests/test_keras.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/integration_tests/test_keras.py:1: in <module>
    from keras import Sequential
E   ModuleNotFoundError: No module named 'keras'
=============================== warnings summary ===============================
tests/integration_tests/test_mlflow.py:7
  /home/runner/work/optuna/optuna/tests/integration_tests/test_mlflow.py:7: FutureWarning: MLflow support for Python 3.6 is deprecated and will be dropped in an upcoming release. At that point, existing Python 3.6 workflows that use MLflow will continue to work without modification, but Python 3.6 users will no longer get access to the latest MLflow features and bugfixes. We recommend that you upgrade to Python 3.7 or newer.
    import mlflow

../../../../../opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/tensorflow/python/autograph/utils/testing.py:[21](https://github.com/optuna/optuna/runs/5091139491?check_suite_focus=true#step:7:21)
  /opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/tensorflow/python/autograph/utils/testing.py:21: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
ERROR tests/integration_tests/test_keras.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
========================= 2 warnings, 1 error in 8.51s =========================
Error: Process completed with exit code 2.
```

**Comment by himkt** [comment]
@nzw0301 found TensorFlow and keras for Python3.6 conflicts with `type_extensions>=3.10.0.0`.
In #3270, we drop TensorFlow support in Python3.6. Please wait a moment until #3270 is merged. :pray:

**Comment by himkt** [comment]
@higucheese #3296 was merged. Please merge or rebase the lates master branch. :pray:

**Review by himkt — APPROVED** [review]
LGTM. Thank you for the contribution!

One minor point: is it better to create `_convert_positional_args.py` in `optuna/study` to align `_test_convert_positional_args.py`? (I'm neutral)

**Comment by hvy** [comment]
> One minor point: is it better to create _convert_positional_args.py in optuna/study to align _test_convert_positional_args.py? (I'm neutral)

That sounds fair. `study.py` is too large already, and the logic is well isolated. Do you have time to continue with this change in this PR, given it's already approved by one reviewer? @higucheese, @himkt 
(This is not a strong opinion, but it'd also be fine to place it under `optuna/` since it currently assumes very little about a study, merely that it's a callable.)

**Comment by himkt** [comment]
The suggestion to move to `optuna/` sounds solid, and I can take over to move the file (if @higucheese is busy for the change). 🙋‍♂️

File: optuna/study/study.py
Comment (hvy) [inline optuna/study/study.py:76]: Ah, we should use a `FutureWarning` here, similar to https://github.com/optuna/optuna/blob/release-v3.0.0-a1/optuna/_deprecated.py.
```suggestion
                warnings.warn(
                    f"{func.__name__}(): Positional arguments are deprecated."
                    " Please give all values as keyword arguments.",
                    FutureWarning
                )
```

File: optuna/study/study.py
Comment (hvy) [inline optuna/study/study.py:68]: How about allowing `func` to change the order of the (keyword only) arguments? That is, relax the assertion to only check that `previous_positional_arg_names` is a subset.

**Comment by hvy** [comment]
> The suggestion to move to optuna/ sounds solid, and I can take over to move the file (if @higucheese is busy for the change). 🙋‍♂️

Thanks a lot. I'm fine either way. What do you prefer @higucheese ?
The current changes by the way LGTM except for the last few comments. PTAL. 🙇 

_On a side note, certain unfamiliar with libraries enforcing keyword arguments may find this change a bit obtrusive. Let's keep #2911 open for feedback._

File: optuna/study/study.py
Comment (higucheese) [reply optuna/study/study.py]: Fix in 81504f6 and introduce `stacklevel` to indicate proper position in stack trace in 7229d24. The `warning_stacklevel` is required when decorators are nested.

File: optuna/study/study.py
Comment (higucheese) [reply optuna/study/study.py]: Sure. Relax the check in f5103d5.

**Comment by higucheese** [comment]
For the suggestion, I am going to do the move by myself if it's no trouble (This PR is not the kind of thing that needs to be merged in a hurry, is it?).

**Comment by higucheese** [comment]
> _On a side note, certain users unfamiliar with libraries enforcing keyword arguments may find this change a bit obtrusive. Let's keep https://github.com/optuna/optuna/issues/2911 open for feedback._

Do we need any comments (or printing info.) to lead users to the issue?

**Comment by himkt** [comment]
> This PR is not the kind of thing that needs to be merged in a hurry, is it?

Thank you so much, no extra rush. 👍

**Comment by higucheese** [comment]
I moved `convert_positional_args` to optuna/ with some additional description. PTAL. 🙇

**Comment by hvy** [comment]
> Do we need any comments (or printing info.) to lead users to the issue?

Sounds good, I think including the URL to #2911 in the warning message could be user friendly. It'll go against moving this decorator to `optuna/`, but I'd image we could create a more general issue and change the URL if this decorator is to be applied to other functions outside the study. What do you think? @higucheese , @himkt

File: optuna/_convert_positional_args.py
Comment (hvy) [inline optuna/_convert_positional_args.py:21]: Sorry but a conventional nit.
```suggestion
    """Convert positional arguments to keyword arguments.
```

File: tests/test_convert_positional_args.py
Comment (hvy) [inline tests/test_convert_positional_args.py:33]: Ditto.
```suggestion
        decorated_func(a=1, b=2, c=3)  # No warning.
```

File: tests/test_convert_positional_args.py
Comment (hvy) [inline tests/test_convert_positional_args.py:82]: Nice checking the content of the message. I'm wondering however if the order (the printed string) can be guaranteed. 🤔

File: tests/test_convert_positional_args.py
Comment (hvy) [inline tests/test_convert_positional_args.py:89]: Definitely not a strong opinion but sounds like the pytest parametrization pattern could be used here, similar to e.g. https://github.com/optuna/optuna/blob/v3.0.0-a1/tests/samplers_tests/tpe_tests/test_parzen_estimator.py#L120.

File: tests/test_convert_positional_args.py
Comment (hvy) [inline tests/test_convert_positional_args.py:22]:
```suggestion
def test_convert_positional_args_user_warning() -> None:
```

File: tests/test_convert_positional_args.py
Comment (hvy) [inline tests/test_convert_positional_args.py:41]:
```suggestion
def test_convert_positional_args_mypy_type_inference() -> None:
```

File: tests/test_convert_positional_args.py
Comment (hvy) [inline tests/test_convert_positional_args.py:69]:
```suggestion
def test_convert_positional_args_invalid_previous_positional_arg_names() -> None:
```

File: tests/test_convert_positional_args.py
Comment (hvy) [inline tests/test_convert_positional_args.py:92]:
```suggestion
def test_convert_positional_args_invalid_positional_args() -> None:
```

File: optuna/study/study.py
Comment (hvy) [inline optuna/study/study.py:1350]: 👍

**Review by hvy — CHANGES_REQUESTED** [review]
Thanks a lot for the quick updates!

File: tests/test_convert_positional_args.py
Comment (higucheese) [reply tests/test_convert_positional_args.py]: Thank you for pointing out. Yes, the order of items in the printed string can vary and that should make this test flaky. I could not notice since this test did not fail a couple of times by luck.
Fixed in 3dd7917.

File: tests/test_convert_positional_args.py
Comment (higucheese) [reply tests/test_convert_positional_args.py]: Thanks! Fixed in 946fdb2 with this change, `user_waning` => `future_warning`.

Comment (hvy) [reply tests/test_convert_positional_args.py]: Good catch, thanks!

File: tests/test_convert_positional_args.py
Comment (hvy) [reply tests/test_convert_positional_args.py]: Thanks!

**Review by hvy — APPROVED** [review]
Thank a lot for the long running PR, LGTM!

Again, I think including the issue url to the warning message could be friendly, it's not a strong opinion though.

**Comment by higucheese** [comment]
Sorry for my late responding. I created #3324 to describe the purpose of these changes and updated the warning comment.

**Comment by himkt** [comment]
Thank you for the update! I'll check it tomorrow.

File: optuna/study/study.py
Comment (himkt) [reply optuna/study/study.py:1350]: [memo]

- warning_stacklevel == 2 (default)

```sh
> python sam.py                                                                                                                                                                                                                    2022-02-22 10:49:18
[I 2022-02-22 10:49:23,077] A new study created in RDB with name: test
/Users/himkt/work/github.com/himkt/optuna/sam.py:5: ExperimentalWarning: copy_study is experimental (supported from v2.8.0). The interface can change in the future.
  optuna.copy_study("test", "sqlite:////tmp/tmp.db", "sqlite:////tmp/tmp2.db")
/Users/himkt/work/github.com/himkt/optuna/optuna/_experimental.py:68: FutureWarning: copy_study(): Please give all values as keyword arguments. See https://github.com/optuna/optuna/issues/3324 for details.
  return func(*args, **kwargs)  # type: ignore
[I 2022-02-22 10:49:23,135] A new study created in RDB with name: test
/Users/himkt/work/github.com/himkt/optuna/optuna/study/study.py:1446: ExperimentalWarning: add_trials is experimental (supported from v2.5.0). The interface can change in the future.
  to_study.add_trials(from_study.get_trials(deepcopy=False))
```

- warning_stacklevel == 3

```sh
> python sam.py                                                                                                                                                                                                                    2022-02-22 10:49:34
[I 2022-02-22 10:49:37,185] A new study created in RDB with name: test
/Users/himkt/work/github.com/himkt/optuna/sam.py:5: ExperimentalWarning: copy_study is experimental (supported from v2.8.0). The interface can change in the future.
  optuna.copy_study("test", "sqlite:////tmp/tmp.db", "sqlite:////tmp/tmp2.db")
/Users/himkt/work/github.com/himkt/optuna/sam.py:5: FutureWarning: copy_study(): Please give all values as keyword arguments. See https://github.com/optuna/optuna/issues/3324 for details.
  optuna.copy_study("test", "sqlite:////tmp/tmp.db", "sqlite:////tmp/tmp2.db")
[I 2022-02-22 10:49:37,249] A new study created in RDB with name: test
/Users/himkt/work/github.com/himkt/optuna/optuna/study/study.py:1447: ExperimentalWarning: add_trials is experimental (supported from v2.5.0). The interface can change in the future.
  to_study.add_trials(from_study.get_trials(deepcopy=False))
```

**Review by himkt — APPROVED** [review]
Thank you so much @higucheese! LGTM.

---

I just found warning message is repeatedly displayed if stacklevel is 2 (it doesn't happen on an interactive environment). I was concerning the problem of the massive warning as in #3301. But I confirm it doesn't affect other places so I think it is acceptable.

- warn_stacklevel == 2 (default)

```sh
> python main.py                                                                                                                                                                                                                   2022-02-22 10:57:33
/Users/himkt/work/github.com/himkt/optuna/main.py:8: FutureWarning: create_study(): Please give all values as keyword arguments. See https://github.com/optuna/optuna/issues/3324 for details.
  study = optuna.create_study("sqlite:////tmp/tmp.db", study_name="test1")
[I 2022-02-22 10:57:35,472] A new study created in RDB with name: test1
/Users/himkt/work/github.com/himkt/optuna/main.py:9: FutureWarning: create_study(): Please give all values as keyword arguments. See https://github.com/optuna/optuna/issues/3324 for details.
  study = optuna.create_study("sqlite:////tmp/tmp.db", study_name="test2")
[I 2022-02-22 10:57:35,492] A new study created in RDB with name: test2
/Users/himkt/work/github.com/himkt/optuna/optuna/trial/_trial.py:169: FutureWarning: UniformDistribution has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use :class:`~optuna.distributions.FloatDistribution` instead.
  distribution = UniformDistribution(low=low, high=high)
/Users/himkt/work/github.com/himkt/optuna/optuna/distributions.py:561: FutureWarning: UniformDistribution has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use :class:`~optuna.distributions.FloatDistribution` instead.
  return cls(**json_dict["attributes"])
[I 2022-02-22 10:57:35,523] Trial 0 finished with value: 5.472289124011194 and parameters: {'v': 5.472289124011194}. Best is trial 0 with value: 5.472289124011194.
[I 2022-02-22 10:57:35,540] Trial 1 finished with value: 1.8239470731739338 and parameters: {'v': 1.8239470731739338}. Best is trial 1 with value: 1.8239470731739338.
[I 2022-02-22 10:57:35,556] Trial 2 finished with value: 3.1063082890049643 and parameters: {'v': 3.1063082890049643}. Best is trial 1 with value: 1.8239470731739338.
```

- warn_stacklevel == 1

```sh
> python main.py                                                                                                                                                                                                                   2022-02-22 10:59:12
/Users/himkt/work/github.com/himkt/optuna/optuna/_convert_positional_args.py:40: FutureWarning: create_study(): Please give all values as keyword arguments. See https://github.com/optuna/optuna/issues/3324 for details.
  warnings.warn(
[I 2022-02-22 10:59:16,198] A new study created in RDB with name: test1
[I 2022-02-22 10:59:16,221] A new study created in RDB with name: test2
/Users/himkt/work/github.com/himkt/optuna/optuna/trial/_trial.py:169: FutureWarning: UniformDistribution has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use :class:`~optuna.distributions.FloatDistribution` instead.
  distribution = UniformDistribution(low=low, high=high)
/Users/himkt/work/github.com/himkt/optuna/optuna/distributions.py:561: FutureWarning: UniformDistribution has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use :class:`~optuna.distributions.FloatDistribution` instead.
  return cls(**json_dict["attributes"])
[I 2022-02-22 10:59:16,259] Trial 0 finished with value: 1.4824433429684192 and parameters: {'v': 1.4824433429684192}. Best is trial 0 with value: 1.4824433429684192.
[I 2022-02-22 10:59:16,278] Trial 1 finished with value: 4.64190930455168 and parameters: {'v': 4.64190930455168}. Best is trial 0 with value: 1.4824433429684192.
[I 2022-02-22 10:59:16,297] Trial 2 finished with value: 0.16982925381998415 and parameters: {'v': 0.16982925381998415}. Best is trial 2 with value: 0.16982925381998415.
```

```diff
diff --git a/optuna/_convert_positional_args.py b/optuna/_convert_positional_args.py
index cfcd6b4b4..f9936e93b 100644
--- a/optuna/_convert_positional_args.py
+++ b/optuna/_convert_positional_args.py
@@ -36,6 +36,7 @@ def convert_positional_args(
         def converter_wrapper(*args: Any, **kwargs: Any) -> _T:

             if len(args) >= 1:
+                warning_stacklevel = 1
                 warnings.warn(
                     f"{func.__name__}(): Please give all values as keyword arguments."
                     " See https://github.com/optuna/optuna/issues/3324 for details.",
```

**Comment by hvy** [comment]
Thanks @himkt for the review, and again @higucheese for the PR.
Not entirely sure but maybe it's because with this deeper stack level, the warning/stack start to differ with the code that you provide. Study names differ.

**Pull request merged**
